### ISSUE TITLE
Diskless resource not created (sometimes)

### ISSUE THREAD

**Issue by chiraganand** [OpenNebula/addon-linstor_un#12]
We are seeing a strange bug while creating diskless resource on the host where VM would run:

```
Mon Jul 20 14:53:40 2020 [Z0][VMM][I]: Command execution fail: cat << EOT | /var/tmp/one/vmm/kvm/deploy '/var/lib/one//datastores/111/2317/deployment.0' 'XXX-host' 2317 XXX-host
Mon Jul 20 14:53:40 2020 [Z0][VMM][I]: error: Failed to create domain from /var/lib/one//datastores/111/2317/deployment.0
Mon Jul 20 14:53:40 2020 [Z0][VMM][I]: error: Cannot access storage file '/var/lib/one//datastores/111/2317/disk.0': No such file or directory
Mon Jul 20 14:53:40 2020 [Z0][VMM][E]: Could not create domain from /var/lib/one//datastores/111/2317/deployment.0
```

We diagnosed that the diskless resource is not getting created on `XXX-host`, hence, the `No such file or directory` error message. In our topology we have separate VM hosts which only run VMs and only contain diskless resources. There are separate storage nodes which do not run any VMs (diskful nodes).

`XXX-host` is a pure KVM host without any local storage, is not part of `BRIDGE_LIST` but is part of diskless storage pool. By default all storage nodes are also part of default diskless pool.

The problem is happening at this part of the code: On a previous version of linstor_un: https://github.com/OpenNebula/addon-linstor_un/blob/b47fd718410073ff1b96332410fc6ccb8f68a5b3/tm/linstor_un/clone#L201-L205. FYI: We are using this version of linstor_un.

In the above code, Linstor, sometimes, ends up creating a diskless resource on one of the storage nodes which do not have VM running capability and not on `XXX-host`. This problem gets fixed if we remove `-s $DISKLESS_POOL` and add `--drbd-diskless` switch to the command. Linstor [documentation](https://www.linbit.com/drbd-user-guide/linstor-guide-1_0-en/#s-drbd_clients) states:
> By using the --drbd-diskless option instead of --storage-pool you can have a permanently diskless DRBD device on a node.

It seems that Linstor selects any node from the diskless pool even if when the host requested is provided. Or is it that this is happening due to some other configuration issue?

**Comment by kvaps** [comment]
Hi, there were couple of changes which fixes this behavior:

First the [`linstor_attach_diskless`](https://github.com/OpenNebula/addon-linstor_un/blob/cd7d19b96de9f07eb58925df587809a4ec0fe3ac/datastore/linstor_un/linstor_utils.sh#L380-L424) function were introduced with the locking mechanism implementation, it solves multiple concurrent copies, and guaranies that disk replica will not be deleted until other clone actions are using it.

Second thing is that issue https://github.com/LINBIT/linstor-server/issues/106#issuecomment-584107852, it was already fixed on the LINSTOR side, but after a while the problem has occurred again. I fixed this in latest commit https://github.com/OpenNebula/addon-linstor_un/commit/ea0ad567997eefa82d6debfb0fca9b04caa535a6

**Comment by chiraganand** [comment]
> Hi, there were couple of changes which fixes this behavior:
> 
> First the [`linstor_attach_diskless`](https://github.com/OpenNebula/addon-linstor_un/blob/cd7d19b96de9f07eb58925df587809a4ec0fe3ac/datastore/linstor_un/linstor_utils.sh#L380-L424) function were introduced with the locking mechanism implementation, it solves multiple concurrent copies, and guaranies that disk replica will not be deleted until other clone actions are using it.

Okay, this is great!

> Second thing is that issue [LINBIT/linstor-server#106 (comment)](https://github.com/LINBIT/linstor-server/issues/106#issuecomment-584107852), it was already fixed on the LINSTOR side, but after a while the problem has occurred again. I fixed this in latest commit [ea0ad56](https://github.com/OpenNebula/addon-linstor_un/commit/ea0ad567997eefa82d6debfb0fca9b04caa535a6)

Yes, looks like your patch should fix the behaviour described in LINBIT/linstor-server#106 for our setup too. Thanks!

**Comment by chiraganand** [comment]
This issue seems to have gone now diskless resources are being created on VM Host. Closing this, will reopen in case the bug hits again. Thanks!
